drivers/spi/spi_ll_stm32.c

```diff
@@ -25,52 +25,152 @@
 #define CONFIG_DATA(cfg)					\
 ((struct spi_stm32_data * const)(cfg)->dev->driver_data)
 
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_transmit(SPI_TypeDef *spi, struct spi_stm32_data *data);
-static void spi_stm32_receive(SPI_TypeDef *spi, struct spi_stm32_data *data);
+#ifdef LL_SPI_SR_UDR
+#define SPI_STM32_ERR_MSK (LL_SPI_SR_UDR | LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | \
+			   LL_SPI_SR_OVR | LL_SPI_SR_FRE)
+#else
+#define SPI_STM32_ERR_MSK (LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | \
+			   LL_SPI_SR_OVR | LL_SPI_SR_FRE)
+#endif
 
-static void spi_stm32_isr(void *arg)
+/* Value to shift out when no application data needs transmitting. */
+#define SPI_STM32_TX_NOP 0x00
+
+static bool spi_stm32_transfer_ongoing(struct spi_stm32_data *data)
 {
-	struct device * const dev = (struct device *) arg;
-	const struct spi_stm32_config *cfg = dev->config->config_info;
-	struct spi_stm32_data *data = dev->driver_data;
-	SPI_TypeDef *spi = cfg->spi;
+	return spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx);
+}
+
+static int spi_stm32_get_err(SPI_TypeDef *spi)
+{
+	u32_t sr = LL_SPI_ReadReg(spi, SR);
+
+	return (int)(sr & SPI_STM32_ERR_MSK);
+}
 
-	if (LL_SPI_IsActiveFlag_TXE(spi) &&
-	    (spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx))) {
-		spi_stm32_transmit(spi, data);
+static inline u8_t spi_stm32_next_tx(struct spi_stm32_data *data)
+{
+	return spi_context_tx_on(&data->ctx) ?
+		*data->ctx.tx_buf : SPI_STM32_TX_NOP;
+}
+
+/* Shift a SPI frame as master. */
+static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
+{
+	u8_t tx_frame;
+	u8_t rx_frame;
+
+	tx_frame = spi_stm32_next_tx(data);
+	while (!LL_SPI_IsActiveFlag_TXE(spi)) {
+		/* NOP */
+	}
+	LL_SPI_TransmitData8(spi, tx_frame);
+	/* The update is ignored if TX is off. */
+	spi_context_update_tx(&data->ctx, 1);
+
+	while (!LL_SPI_IsActiveFlag_RXNE(spi)) {
+		/* NOP */
+	}
+	rx_frame = LL_SPI_ReceiveData8(spi);
+	if (spi_context_rx_on(&data->ctx)) {
+		*data->ctx.rx_buf = rx_frame;
+		spi_context_update_rx(&data->ctx, 1);
+	}
+}
```

Comment:
I don't get the true difference apart that in the master shift you consider you want to push a byte and then get a byte, but why is it different for slave and why can't we leave the previous way for both ?